src/l1_fetcher.rs

```diff
@@ -183,20 +185,20 @@ impl L1Fetcher {
         Ok(())
     }
 
-    async fn get_logs(&self, filter: &Filter) -> Result<Vec<Log>> {
+    async fn retry_call<T, Fut>(&self, callback: impl Fn() -> Fut) -> Result<T>
+    where
+        Fut: Future<Output = Result<T>>,
+    {
```

Comment:
Nice, this looks perfect!

Reply:
Or almost, you should probably specify the exact error the closure returns, since we know that part: 
```suggestion
    async fn retry_call<T, Fut>(&self, callback: impl Fn() -> Fut) -> Result<T>
    where
        Fut: Future<Output = std::result::Result<T, ProviderError>>,
    {
```